### PR TITLE
CAS: Clean up printing code for CASID

### PR DESCRIPTION
**Pull request by dexonsmith**
Now that a CASID points at its context, it shouldn't be so hard to
print, dump, or convert to a string.

- Change the printing method in CASIDContext to `protected` to ensure
  it's only called from CASID.
- Replace Error paths with an assertion.
- Remove other CASIDContext APIs, and replace them with APIs on CASID
  that are easier to use.
- Add `CASID::dump()` for use in LLDB.

### CODE REVIEW

**Comment by dexonsmith** [comment]
@swift-ci test

**Comment by dexonsmith** [comment]
Looks like macOS is offline:
https://ci.swift.org/job/apple-llvm-project-pr-macos/305/console

Merging since Linux passed.

File: clang/tools/clang-scan-deps/ClangScanDeps.cpp [hunk 1]
```diff
@@ -288,9 +288,7 @@ handleTreeDependencyToolResult(llvm::cas::CASDB &CAS, const std::string &Input,
     return true;
   }
   OS.applyLocked([&](llvm::raw_ostream &OS) {
-    OS << "tree ";
-    llvm::cantFail(CAS.printCASID(OS, *MaybeTree));
-    OS << " for '" << Input << "'\n";
+    OS << "tree " << *MaybeTree << " for '" << Input << "'\n";
```
Comment (cachemeifyoucan) [inline clang/tools/clang-scan-deps/ClangScanDeps.cpp:291]: How does this work? TreeRef doesn't have << operator. This breaks auto merger.

Comment (dexonsmith) [reply clang/tools/clang-scan-deps/ClangScanDeps.cpp:291]: Hmm... how did Linux tests pass then?

Fix is `MaybeTree.getID()`. Will push in a moment i fyou haven't already fixed it.

Comment (dexonsmith) [reply clang/tools/clang-scan-deps/ClangScanDeps.cpp:291]: Pushed in 007f2611a4ae702885e47eff4f5ddcef4748901c. Also fixes `llc` build which had (a) a related issue and (b) an unclosed parenthesis?

Does the Linux build not actually do anything?

**Pull request merged**